src/compiler/binder.ts

```diff
@@ -1446,27 +1446,114 @@ namespace ts {
         }
 
         function bindBinaryExpressionFlow(node: BinaryExpression) {
-            const operator = node.operatorToken.kind;
-            if (operator === SyntaxKind.AmpersandAmpersandToken || operator === SyntaxKind.BarBarToken || operator === SyntaxKind.QuestionQuestionToken) {
-                if (isTopLevelLogicalExpression(node)) {
-                    const postExpressionLabel = createBranchLabel();
-                    bindLogicalExpression(node, postExpressionLabel, postExpressionLabel);
-                    currentFlow = finishFlowLabel(postExpressionLabel);
-                }
-                else {
-                    bindLogicalExpression(node, currentTrueTarget!, currentFalseTarget!);
+            const work: [BinaryExpression, number, boolean?, Node?, number?][] = [[node, 0, undefined, undefined, undefined]];
+            while (work.length) {
+                const res = work[work.length - 1];
+                node = res[0];
+                switch (res[1]) {
+                    case -1: {
+                        // The -1 state is used only when recuring, to emulate the work that `bind` does before
+                        // reaching `bindChildren`. A normal call to `bindBinaryExpressionFlow` will already have done this work.
+                        node.parent = parent;
+                        const saveInStrictMode = inStrictMode;
+                        bindWorker(node);
+                        const saveParent = parent;
+                        parent = node;
+
+                        // While this next part does the work of `bindChildren` before it descends into `bindChildrenWorker`
+                        // and uses `res[4]` to queue up the work that needs to be done once the node is bound.
+                        if (skipTransformFlagAggregation) {
+                            /// do nothing extra
+                        }
+                        else if (node.transformFlags & TransformFlags.HasComputedFlags) {
+                            skipTransformFlagAggregation = true;
+                            res[4] = -1;
+                        }
+                        else {
+                            const savedSubtreeTransformFlags = subtreeTransformFlags;
+                            subtreeTransformFlags = 0;
+                            res[4] = savedSubtreeTransformFlags;
+                        }
+
+
+                        res[2] = saveInStrictMode;

```

Comment:
I'd recommend either a const enum for field names in the tuple, or helper functions to get/set (if you are using parallel arrays), i.e.:

```ts
interface BindStack {
  frame: number;
  exprStack: BinaryExpression[];
  stateStack: ...;
  inStrictModeStack: (boolean | undefined)[];
  ...
}
function getInStrictMode(stack: BindStack) {
  return stack.inStrictModeStack[stack.frame];
}
function setInStrictMode(stack: BindStack, value?: boolean) {
  stack.inStrictModeStack[stack.frame] = value;
}

...

setInStrictMode(stack, saveInStrictMode);
```